### ISSUE TITLE
[SB2] Lid Close ACPI Error

### ISSUE THREAD

**Issue by versipellis** [jakeday/linux-surface#335]
Latest kernel, shutting the lid throws these two lines repeating, along with the hissing:
```
surfacegen5_acpi_san MSHW0091:00: ETWL (0x01, 0x02): RQST(17, 13, 0) error: 5
surfacegen5_acpi_san MSHW0091:00: san_rqst: failed with error -1
```

On one occasion while testing the lid close crashing, it threw an error after this page about CPU core temps reaching 58C and shutting down as a result.

**Comment by qzed** [comment]
This is a known issue. It's caused by the device querying the lid-state even though the embedded controller is still suspended. Due to the way this is implemented in the DSDT/ACPI, this check repeats until it succeeds, which it will only once the controller has resumed, causing the CPU to clock up. This should however be only for a short period, and 58°C shouldn't be a critical temperature at all (also the kernel should start to down-clock before a shutdown).

Unfortunately I haven't found a way to delay the lid-state check until the controller has been resumed, as the check itself is initiated via ACPI.

**Comment by versipellis** [comment]
Gotcha, have any temporary solutions in the meantime? Disable lid-close actions?

**Comment by qzed** [comment]
You could do that, or switch to the powersave governor (that should prevent overheating). I'll try to figure out some solution to stop it shutting down, it _really_ shouldn't do that. I think the least intrusive way to fix this is to detect the query and rate-limit it.

**Comment by versipellis** [comment]
Like, permanently setting the CPU power management to the powersave governor, or setting lid-close to do that?

**Comment by qzed** [comment]
Either you can do that manually or automatically prior to lid close/suspend, or permanently if you don't need the extra performance, which it will impact (as a temporary workaround of course, I'm working on this right now). The choice is up to you.

The thing is: It really shouldn't shut down, even if you're in performance mode, sure its constantly running on one core, but only for a fairly short time. It should down-clock way before that. Have you ever gotten it to shut down with any other workload?

**Comment by versipellis** [comment]
It threw the temperature error when I had a bunch of firefox tabs open.
With a clean boot, it happens with no load (besides gnome) on lid-close but
no temperature error. How would you do it prior to lid close/suspend,
rather than _on_ lid close/suspend? (Also, suspend and hibernate don't
work, but that's a separate issue I've already opened :P )

>

**Comment by qzed** [comment]
I think you could hook into ACPI events. Also what do you mean by "suspend is not working", This particular error should only occur when you resume from suspend.

**Comment by qzed** [comment]
PR #336 should fix the overheating issue by delaying subsequent base-state queries (required for the lid-state). Fixing the underlying issue, namely that the lid state (and thus base-state) is queried before the EC is resumed is probably impossible without patching the DSDT itself.

To expand on that a bit: It seems like MS has had the same problem, due to which they chose that in case the EC is suspended, the ACPI code re-queries the state until the EC has resumed. With the fix applied, it will still emit warnings, but the issue of overheating is fixed as the base-state queries are then essentially rate-limited. In this case, the warnings can be ignored, as there are no other implications.

**Comment by versipellis** [comment]
@qzed Was this resolved in the latest kernel release? Just patched up and it appears to still be happening.

**Comment by qzed** [comment]
@versipellis It will still emit warnings, it won't overheat though.

**Comment by versipellis** [comment]
Gotcha. But the lid-close shutdown matter still happens, so I'm not sure why this issue was closed.

**Comment by qzed** [comment]
So you're saying it still overheats? That should be impossible now.

**Comment by versipellis** [comment]
It just shuts down. It doesn't overheat. Tested on cold boot.

**Comment by qzed** [comment]
Okay, that might be a configuration issue then. Anything interesting in the logs? Have you replaced suspend with hibernate?

**Comment by versipellis** [comment]
`syslog`:
```
Jan 13 22:23:32 Hyde kernel: [ 4684.556113] surfacegen5_acpi_ssh serial0-0: rqst: embedded controller is suspended
Jan 13 22:23:32 Hyde kernel: [ 4685.569866] surfacegen5_acpi_ssh serial0-0: rqst: embedded controller is suspended
Jan 13 22:23:32 Hyde kernel: [ 4686.592518] surfacegen5_acpi_san MSHW0091:00: unsupported request: RQSG(0x13, 0x02, 0x00)
Jan 13 22:23:32 Hyde kernel: [ 4686.594576] surfacegen5_acpi_san MSHW0091:00: ETWL(0x01, 0x02): RQST(19, 2, 0) error: 6
Jan 13 22:23:32 Hyde kernel: [ 4686.594735] ACPI: button: The lid device is not compliant to SW_LID.
Jan 13 22:23:32 Hyde kernel: [ 4686.599094] surfacegen5_acpi_ssh serial0-0: rqst: embedded controller is suspended
Jan 13 22:23:32 Hyde kernel: [ 4686.656233] [drm] GuC: Loaded firmware i915/kbl_guc_ver9_39.bin (version 9.39)
Jan 13 22:23:32 Hyde kernel: [ 4686.656318] i915 0000:00:02.0: GuC firmware version 9.39
Jan 13 22:23:32 Hyde kernel: [ 4686.656319] i915 0000:00:02.0: GuC submission enabled
Jan 13 22:23:32 Hyde kernel: [ 4686.656320] i915 0000:00:02.0: HuC disabled
Jan 13 22:23:32 Hyde kernel: [ 4686.868228] [drm] Reducing the compressed framebuffer size. This may lead to less power savings than a non-reduced-size. Try to increase stolen memory size if available in BIOS.
Jan 13 22:23:32 Hyde kernel: [ 4687.617362] surfacegen5_acpi_ssh serial0-0: rqst: embedded controller is suspended
Jan 13 22:23:32 Hyde kernel: [ 4688.641379] surfacegen5_acpi_ssh serial0-0: rqst: embedded controller is suspended
Jan 13 22:23:32 Hyde kernel: [ 4689.680711] acpi LNXPOWER:06: Turning OFF
Jan 13 22:23:32 Hyde kernel: [ 4689.690099] PM: Using 3 thread(s) for compression
Jan 13 22:23:32 Hyde kernel: [ 4689.690102] PM: Compressing and saving image data (1175133 pages)...
Jan 13 22:23:32 Hyde kernel: [ 4689.690148] PM: Image saving progress:   0%
Jan 13 22:23:32 Hyde kernel: [ 4690.163051] PM: Image saving progress:  10%
Jan 13 22:23:32 Hyde kernel: [ 4690.529374] PM: Image saving progress:  20%
Jan 13 22:23:32 Hyde kernel: [ 4690.937611] PM: Image saving progress:  30%
Jan 13 22:23:32 Hyde kernel: [ 4691.462698] PM: Image saving progress:  40%
Jan 13 22:23:32 Hyde kernel: [ 4692.068181] PM: Image saving progress:  50%
Jan 13 22:23:32 Hyde kernel: [ 4692.510845] PM: Image saving progress:  60%
Jan 13 22:23:32 Hyde kernel: [ 4692.862896] PM: Image saving progress:  70%
Jan 13 22:23:32 Hyde kernel: [ 4692.937446] PM: Wrote 4700532 kbytes in 3.24 seconds (1450.78 MB/s)
Jan 13 22:23:32 Hyde kernel: [ 4693.047142] PM: Basic memory bitmaps freed
Jan 13 22:23:32 Hyde kernel: [ 4693.047144] OOM killer enabled.
Jan 13 22:23:32 Hyde kernel: [ 4693.047145] Restarting tasks ... done.
Jan 13 22:23:32 Hyde acpid: client 1715[0:0] has disconnected
Jan 13 22:23:32 Hyde whoopsie[1587]: [22:23:32] Cannot reach: https://daisy.ubuntu.com
Jan 13 22:23:32 Hyde whoopsie[1587]: [22:23:32] offline
Jan 13 22:23:32 Hyde rtkit-daemon[1431]: The canary thread is apparently starving. Taking action.
Jan 13 22:23:32 Hyde rtkit-daemon[1431]: Demoting known real-time threads.
Jan 13 22:23:32 Hyde rtkit-daemon[1431]: Successfully demoted thread 1920 of process 1918 (n/a).
Jan 13 22:23:32 Hyde rtkit-daemon[1431]: Successfully demoted thread 1919 of process 1918 (n/a).
Jan 13 22:23:32 Hyde rtkit-daemon[1431]: Successfully demoted thread 1918 of process 1918 (n/a).
Jan 13 22:23:32 Hyde rtkit-daemon[1431]: Demoted 3 threads.
Jan 13 22:23:32 Hyde kernel: [ 4693.065552] thermal thermal_zone4: critical temperature reached (64 C), shutting down
Jan 13 22:23:32 Hyde systemd[1]: Stopping Manage, Install and Generate Color Profiles...
Jan 13 22:23:32 Hyde kernel: [ 4693.078251] PM: hibernation exit
Jan 13 22:23:32 Hyde io.snapcraft.Settings[1723]: Exiting on terminated.
```

**Comment by versipellis** [comment]
I ran another test, making sure that the laptop was completely cool. I get the same behavior as before, hissing static followed by a black screen, and then this ACPI error:

`surfacegen5_acpi_san MSHW0091:00: ETWL(0x01, 0x02): RQST(19, 2, 0) error: 6`

No thermal errors this time.

**Comment by qzed** [comment]
```
surfacegen5_acpi_san MSHW0091:00: ETWL(0x01, 0x02): RQST(19, 2, 0) error: 6
```
That's a different error, before it was `RQST(17, ...)`. `RQST(19, ...)` is a request related to graphics, but that's a wontfix for now as I don't know what exactly that request should achieve. It's currently unsupported and thus silently ignored as indicated in the line before:
```
surfacegen5_acpi_san MSHW0091:00: unsupported request: RQSG(0x13, 0x02, 0x00)
```

Also the warnings caused by the lid/base state request are now correctly 1s apart:
```
[ 4684.556113] surfacegen5_acpi_ssh serial0-0: rqst: embedded controller is suspended
[ 4685.569866] surfacegen5_acpi_ssh serial0-0: rqst: embedded controller is suspended
```

It looks like you're using hibernate, I'm a bit surprised that resuming from that works with opening the lid state, or are you using a hybrid mode? What's also weird is that 
```
[ 4689.690102] PM: Compressing and saving image data (1175133 pages)...
```
comes after you've already resumed. This kind of leads me to think that you're triggering suspend _and_ hibernate at the same time.

The next thing is:
```
thermal thermal_zone4: critical temperature reached (64 C), shutting down
```
that's not the CPU, at least for me. Can you check what
```
cat /sys/class/thermal/thermal_zone4/type
```
gives you? For me this is `GEN5`, indicating the battery temperature, which would be quite odd.

**Comment by versipellis** [comment]
`xxx@Hyde:~$ cat /sys/class/thermal/thermal_zone4/type
GEN6`

This is all coming off of a clean install of Ubuntu and Jakeday kernel, using the options specified in the setup.sh to replace suspend with hibernate.

I wonder if the graphics is related to using the nvidia 390.87 drivers. I didn't have any issues with this prior to the latest jakeday kernel.

**Comment by qzed** [comment]
> I wonder if the graphics is related to using the nvidia 390.87 drivers.

`GEN6` is dGPU temperature. Let's focus on the shutdown issue first, that might mess some stuff up, which might cause the dGPU to misbehave.

>  using the options specified in the setup.sh to replace suspend with hibernate

in that case it should hibernate when you close the lid. I'm not sure why it also enters suspend. Do you have something else installed that could trigger this, like acpid? What happens if you change the lid close action in the settings?

**Comment by versipellis** [comment]
Ha, I found the culprit of the dual hibernate-suspend issue: Gnome tweak tools had an option for on-lid close suspend. Disabled that and the problem stops occurring. Cheers!

**Comment by qzed** [comment]
Nice! Is there still any issue with the dGPU? If so I think you should open a new issue.

**Comment by versipellis** [comment]
Doesn't appear to be so. Nothing crashed, so I didn't bother checking the
logs. I can take a look at that tomorrow to see if there's any issues in
the background though.

**Comment by qzed** [comment]
Okay, I guess this was a quirk from the parallel suspend/hibernate then.
